File: dailyQuestion/2020-07/07-03/golang/solution_recurse.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	n := len(nums)
	if n == 0 {
		return nil
	}
	mid := n >> 1
	root := &TreeNode{Val: nums[mid]}
	root.Left = sortedArrayToBST(nums[:mid])
	root.Right = sortedArrayToBST(nums[mid+1:])
	return root
}

func main() {
	result := sortedArrayToBST([]int{-10, -3, 0, 5, 9})
	fmt.Println(result)
}

File: dailyQuestion/2020/2020-07/07-07/python/solution_recurse.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def hasPathSum(root: TreeNode, sum: int) -> bool:
    if not root:
        return False
    if not root.left and not root.right:
        return root.val == sum
    return hasPathSum(root.left, sum - root.val) or hasPathSum(root.right, sum - root.val)

if __name__ == "__main__" :
    node = TreeNode(5)
    node.left = TreeNode(4)
    node.left.left = TreeNode(11)
    node.left.left.left = TreeNode(7)
    node.left.left.right = TreeNode(2)
    node.right = TreeNode(8)
    node.right.right = TreeNode(4)
    node.right.right.right = TreeNode(1)
    result = hasPathSum(node,22)
    print(result)
